lib/libc/sys/sendfile.2
.\" Copyright (c) 2003, David G. Lawrence
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice unmodified, this list of conditions, and the following
.\"    disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" $FreeBSD: head/lib/libc/sys/sendfile.2 293439 2016-01-08 20:34:57Z glebius $
.\"
.Dd January 7, 2016
.Dt SENDFILE 2
.Os
.Sh NAME
.Nm sendfile
.Nd send a file to a socket
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In sys/types.h
.In sys/socket.h
.In sys/uio.h
.Ft int
.Fo sendfile
.Fa "int fd" "int s" "off_t offset" "size_t nbytes"
.Fa "struct sf_hdtr *hdtr" "off_t *sbytes" "int flags"
.Fc
.Sh DESCRIPTION
The
.Fn sendfile
system call
sends a regular file or shared memory object specified by descriptor
.Fa fd
out a stream socket specified by descriptor
.Fa s .
.Pp
The
.Fa offset
argument specifies where to begin in the file.
Should
.Fa offset
fall beyond the end of file, the system will return
success and report 0 bytes sent as described below.
The
.Fa nbytes
argument specifies how many bytes of the file should be sent, with 0 having the special
meaning of send until the end of file has been reached.
.Pp
An optional header and/or trailer can be sent before and after the file data by specifying
a pointer to a
.Vt "struct sf_hdtr" ,
which has the following structure:
.Pp
.Bd -literal -offset indent -compact
struct sf_hdtr {
	struct iovec *headers;	/* pointer to header iovecs */
	int hdr_cnt;		/* number of header iovecs */
	struct iovec *trailers;	/* pointer to trailer iovecs */
	int trl_cnt;		/* number of trailer iovecs */
};
.Ed
.Pp
The
.Fa headers
and
.Fa trailers
pointers, if
.Pf non- Dv NULL ,
point to arrays of
.Vt "struct iovec"
structures.
See the
.Fn writev
system call for information on the iovec structure.
The number of iovecs in these
arrays is specified by
.Fa hdr_cnt
and
.Fa trl_cnt .
.Pp
If
.Pf non- Dv NULL ,
the system will write the total number of bytes sent on the socket to the
variable pointed to by
.Fa sbytes .
.Pp
The least significant 16 bits of
.Fa flags
argument is a bitmap of these values:
.Bl -tag -offset indent
.It Dv SF_NODISKIO
This flag causes
.Nm
to return
.Er EBUSY
instead of blocking when a busy page is encountered.
This rare situation can happen if some other process is now working
with the same region of the file.
It is advised to retry the operation after a short period.
.Pp
Note that in older
.Fx
versions the
.Dv SF_NODISKIO
had slightly different notion.
The flag prevented
.Nm
to run I/O operations in case if an invalid (not cached) page is encountered,
thus avoiding blocking on I/O.
Starting with
.Fx 11
.Nm
sending files off the
.Xr ffs 7
filesystem doesn't block on I/O
(see 
.Sx IMPLEMENTATION NOTES
), so the condition no longer applies.
However, it is safe if an application utilizes
.Dv SF_NODISKIO
and on
.Er EBUSY
performs the same action as it did in
older
.Fx
versions, e.g.
.Xr aio_read 2,
.Xr read 2
or
.Nm
in a different context.
.It Dv SF_NOCACHE
The data sent to socket will not be cached by the virtual memory system,
and will be freed directly to the pool of free pages.
.It Dv SF_SYNC
.Nm
sleeps until the network stack no longer references the VM pages
of the file, making subsequent modifications to it safe.
Please note that this is not a guarantee that the data has actually
been sent.
.El
.Pp
The most significant 16 bits of
.Fa flags
specify amount of pages that
.Nm
may read ahead when reading the file.
A macro
.Fn SF_FLAGS
is provided to combine readahead amount and flags.
Example shows specifing readahead of 16 pages and
.Dv SF_NOCACHE
flag:
.Pp
.Bd -literal -offset indent -compact
	SF_FLAGS(16, SF_NOCACHE)
.Ed
.Pp
When using a socket marked for non-blocking I/O,
.Fn sendfile
may send fewer bytes than requested.
In this case, the number of bytes successfully
written is returned in
.Fa *sbytes
(if specified),
and the error
.Er EAGAIN
is returned.
.Sh IMPLEMENTATION NOTES
The
.Fx
implementation of
.Fn sendfile
doesn't block on disk I/O when it sends a file off the
.Xr ffs 7
filesystem.
The syscall returns success before the actual I/O completes, and data
is put into the socket later unattended.
However, the order of data in the socket is preserved, so it is safe
to do further writes to the socket.
.Pp
The
.Fx
implementation of
.Fn sendfile
is "zero-copy", meaning that it has been optimized so that copying of the file data is avoided.
.Sh TUNING
On some architectures, this system call internally uses a special
.Fn sendfile
buffer
.Pq Vt "struct sf_buf"
to handle sending file data to the client.
If the sending socket is
blocking, and there are not enough
.Fn sendfile
buffers available,
.Fn sendfile
will block and report a state of
.Dq Li sfbufa .
If the sending socket is non-blocking and there are not enough
.Fn sendfile
buffers available, the call will block and wait for the
necessary buffers to become available before finishing the call.
.Pp
The number of
.Vt sf_buf Ns 's
allocated should be proportional to the number of nmbclusters used to
send data to a client via
.Fn sendfile .
Tune accordingly to avoid blocking!
Busy installations that make extensive use of
.Fn sendfile
may want to increase these values to be inline with their
.Va kern.ipc.nmbclusters
(see
.Xr tuning 7
for details).
.Pp
The number of
.Fn sendfile
buffers available is determined at boot time by either the
.Va kern.ipc.nsfbufs
.Xr loader.conf 5
variable or the
.Dv NSFBUFS
kernel configuration tunable.
The number of
.Fn sendfile
buffers scales with
.Va kern.maxusers .
The
.Va kern.ipc.nsfbufsused
and
.Va kern.ipc.nsfbufspeak
read-only
.Xr sysctl 8
variables show current and peak
.Fn sendfile
buffers usage respectively.
These values may also be viewed through
.Nm netstat Fl m .
.Pp
If a value of zero is reported for
.Va kern.ipc.nsfbufs ,
your architecture does not need to use
.Fn sendfile
buffers because their task can be efficiently performed
by the generic virtual memory structures.
.Sh RETURN VALUES
.Rv -std sendfile
.Sh ERRORS
.Bl -tag -width Er
.It Bq Er EAGAIN
The socket is marked for non-blocking I/O and not all data was sent due to
the socket buffer being filled.
If specified, the number of bytes successfully sent will be returned in
.Fa *sbytes .
.It Bq Er EBADF
The
.Fa fd
argument
is not a valid file descriptor.
.It Bq Er EBADF
The
.Fa s
argument
is not a valid socket descriptor.
.It Bq Er EBUSY
A busy page was encountered and
.Dv SF_NODISKIO
had been specified.
Partial data may have been sent.
.It Bq Er EFAULT
An invalid address was specified for an argument.
.It Bq Er EINTR
A signal interrupted
.Fn sendfile
before it could be completed.
If specified, the number
of bytes successfully sent will be returned in
.Fa *sbytes .
.It Bq Er EINVAL
The
.Fa fd
argument
is not a regular file.
.It Bq Er EINVAL
The
.Fa s
argument
is not a SOCK_STREAM type socket.
.It Bq Er EINVAL
The
.Fa offset
argument
is negative.
.It Bq Er EIO
An error occurred while reading from
.Fa fd .
.It Bq Er ENOBUFS
The system was unable to allocate an internal buffer.
.It Bq Er ENOTCONN
The
.Fa s
argument
points to an unconnected socket.
.It Bq Er ENOTSOCK
The
.Fa s
argument
is not a socket.
.It Bq Er EOPNOTSUPP
The file system for descriptor
.Fa fd
does not support
.Fn sendfile .
.It Bq Er EPIPE
The socket peer has closed the connection.
.El
.Sh SEE ALSO
.Xr netstat 1 ,
.Xr open 2 ,
.Xr send 2 ,
.Xr socket 2 ,
.Xr writev 2 ,
.Xr tuning 7
.Rs
.%A K. Elmeleegy
.%A A. Chanda
.%A A. L. Cox
.%A W. Zwaenepoel
.%T A Portable Kernel Abstraction for Low-Overhead Ephemeral Mapping Management
.%J The Proceedings of the 2005 USENIX Annual Technical Conference
.%P pp 223-236
.%D 2005
.Re
.Sh HISTORY
The
.Fn sendfile
system call
first appeared in
.Fx 3.0 .
This manual page first appeared in
.Fx 3.1 .
In
.Fx 10
support for sending shared memory descriptors had been introduced.
In
.Fx 11
a non-blocking implementation had been introduced.
.Sh AUTHORS
The initial implementation of
.Fn sendfile
system call
and this manual page were written by
.An David G. Lawrence Aq Mt dg@dglawrence.com .
The
.Fx 11
implementation was written by
.An Gleb Smirnoff Aq Mt glebius@FreeBSD.org .


tools/regression/sockets/sendfile/sendfile.c
/*-
 * Copyright (c) 2006 Robert N. M. Watson
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * $FreeBSD: head/tools/regression/sockets/sendfile/sendfile.c 294126 2016-01-16 02:15:13Z ngie $
 */

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/stat.h>
#include <sys/wait.h>

#include <netinet/in.h>

#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <limits.h>
#include <md5.h>
#include <signal.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

/*
 * Simple regression test for sendfile.  Creates a file sized at four pages
 * and then proceeds to send it over a series of sockets, exercising a number
 * of cases and performing limited validation.
 */

#define FAIL(msg)	{printf("# %s\n", msg); \
			return (-1);}

#define FAIL_ERR(msg)	{printf("# %s: %s\n", msg, strerror(errno)); \
			return (-1);}

#define	TEST_PORT	5678
#define	TEST_MAGIC	0x4440f7bb
#define	TEST_PAGES	4
#define	TEST_SECONDS	30

struct test_header {
	uint32_t	th_magic;
	uint32_t	th_header_length;
	uint32_t	th_offset;
	uint32_t	th_length;
	char		th_md5[33];
};

struct sendfile_test {
	uint32_t	hdr_length;
	uint32_t	offset;
	uint32_t	length;
	uint32_t	file_size;
};

static int	file_fd;
static char	path[PATH_MAX];
static int	listen_socket;
static int	accept_socket;

static int test_th(struct test_header *th, uint32_t *header_length,
		uint32_t *offset, uint32_t *length);
static void signal_alarm(int signum);
static void setup_alarm(int seconds);
static void cancel_alarm(void);
static int receive_test(void);
static void run_child(void);
static int new_test_socket(int *connect_socket);
static void init_th(struct test_header *th, uint32_t header_length, 
		uint32_t offset, uint32_t length);
static int send_test(int connect_socket, struct sendfile_test);
static int write_test_file(size_t file_size);
static void run_parent(void);
static void cleanup(void);


static int
test_th(struct test_header *th, uint32_t *header_length, uint32_t *offset, 
		uint32_t *length)
{

	if (th->th_magic != htonl(TEST_MAGIC))
		FAIL("magic number not found in header")
	*header_length = ntohl(th->th_header_length);
	*offset = ntohl(th->th_offset);
	*length = ntohl(th->th_length);
	return (0);
}

static void
signal_alarm(int signum)
{
	(void)signum;

	printf("# test timeout\n");

	if (accept_socket > 0)
		close(accept_socket);
	if (listen_socket > 0)
		close(listen_socket);

	_exit(-1);
}

static void
setup_alarm(int seconds)
{
	struct itimerval itv;
	bzero(&itv, sizeof(itv));
	(void)seconds;
	itv.it_value.tv_sec = seconds;

	signal(SIGALRM, signal_alarm);
	setitimer(ITIMER_REAL, &itv, NULL);
}

static void
cancel_alarm(void)
{
	struct itimerval itv;
	bzero(&itv, sizeof(itv));
	setitimer(ITIMER_REAL, &itv, NULL);
}

static int
receive_test(void)
{
	uint32_t header_length, offset, length, counter;
	struct test_header th;
	ssize_t len;
	char buf[10240];
	MD5_CTX md5ctx;
	char *rxmd5;

	len = read(accept_socket, &th, sizeof(th));
	if (len < 0 || (size_t)len < sizeof(th))
		FAIL_ERR("read")

	if (test_th(&th, &header_length, &offset, &length) != 0)
		return (-1);

	MD5Init(&md5ctx);

	counter = 0;
	while (1) {
		len = read(accept_socket, buf, sizeof(buf));
		if (len < 0 || len == 0)
			break;
		counter += len;
		MD5Update(&md5ctx, buf, len);
	}

	rxmd5 = MD5End(&md5ctx, NULL);
	
	if ((counter != header_length+length) || 
			memcmp(th.th_md5, rxmd5, 33) != 0)
		FAIL("receive length mismatch")

	free(rxmd5);
	return (0);
}

static void
run_child(void)
{
	struct sockaddr_in sin;
	int rc = 0;

	listen_socket = socket(PF_INET, SOCK_STREAM, 0);
	if (listen_socket < 0) {
		printf("# socket: %s\n", strerror(errno));
		rc = -1;
	}

	if (!rc) {
		bzero(&sin, sizeof(sin));
		sin.sin_len = sizeof(sin);
		sin.sin_family = AF_INET;
		sin.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
		sin.sin_port = htons(TEST_PORT);

		if (bind(listen_socket, (struct sockaddr *)&sin, sizeof(sin)) < 0) {
			printf("# bind: %s\n", strerror(errno));
			rc = -1;
		}
	}

	if (!rc && listen(listen_socket, -1) < 0) {
		printf("# listen: %s\n", strerror(errno));
		rc = -1;
	}

	if (!rc) {
		accept_socket = accept(listen_socket, NULL, NULL);	
		setup_alarm(TEST_SECONDS);
		if (receive_test() != 0)
			rc = -1;
	}

	cancel_alarm();
	if (accept_socket > 0)
		close(accept_socket);
	if (listen_socket > 0)
		close(listen_socket);

	_exit(rc);
}

static int
new_test_socket(int *connect_socket)
{
	struct sockaddr_in sin;
	int rc = 0;

	*connect_socket = socket(PF_INET, SOCK_STREAM, 0);
	if (*connect_socket < 0)
		FAIL_ERR("socket")

	bzero(&sin, sizeof(sin));
	sin.sin_len = sizeof(sin);
	sin.sin_family = AF_INET;
	sin.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
	sin.sin_port = htons(TEST_PORT);

	if (connect(*connect_socket, (struct sockaddr *)&sin, sizeof(sin)) < 0)
		FAIL_ERR("connect")

	return (rc);
}

static void
init_th(struct test_header *th, uint32_t header_length, uint32_t offset, 
		uint32_t length)
{
	bzero(th, sizeof(*th));
	th->th_magic = htonl(TEST_MAGIC);
	th->th_header_length = htonl(header_length);
	th->th_offset = htonl(offset);
	th->th_length = htonl(length);

	MD5FileChunk(path, th->th_md5, offset, length);
}

static int
send_test(int connect_socket, struct sendfile_test test)
{
	struct test_header th;
	struct sf_hdtr hdtr, *hdtrp;
	struct iovec headers;
	char *header;
	ssize_t len;
	int length;
	off_t off;

	len = lseek(file_fd, 0, SEEK_SET);
	if (len != 0)
		FAIL_ERR("lseek")

	struct stat st;
	if (fstat(file_fd, &st) < 0)
		FAIL_ERR("fstat")
	length = st.st_size - test.offset;
	if (test.length > 0 && test.length < (uint32_t)length)
		length = test.length;

	init_th(&th, test.hdr_length, test.offset, length);

	len = write(connect_socket, &th, sizeof(th));
	if (len != sizeof(th))
		return (-1);

	if (test.hdr_length != 0) {
		header = malloc(test.hdr_length);
		if (header == NULL)
			FAIL_ERR("malloc")

		hdtrp = &hdtr;
		bzero(&headers, sizeof(headers));
		headers.iov_base = header;
		headers.iov_len = test.hdr_length;
		bzero(&hdtr, sizeof(hdtr));
		hdtr.headers = &headers;
		hdtr.hdr_cnt = 1;
		hdtr.trailers = NULL;
		hdtr.trl_cnt = 0;
	} else {
		hdtrp = NULL;
		header = NULL;
	}

	if (sendfile(file_fd, connect_socket, test.offset, test.length, 
				hdtrp, &off, 0) < 0) {
		if (header != NULL)
			free(header);
		FAIL_ERR("sendfile")
	}

	if (length == 0) {
		struct stat sb;

		if (fstat(file_fd, &sb) == 0)
			length = sb.st_size - test.offset;
	}

	if (header != NULL)
		free(header);

	if (off != length)
		FAIL("offset != length")

	return (0);
}

static int
write_test_file(size_t file_size)
{
	char *page_buffer;
	ssize_t len;
	static size_t current_file_size = 0;

	if (file_size == current_file_size)
		return (0);
	else if (file_size < current_file_size) {
		if (ftruncate(file_fd, file_size) != 0)
			FAIL_ERR("ftruncate");
		current_file_size = file_size;
		return (0);
	}

	page_buffer = malloc(file_size);
	if (page_buffer == NULL)
		FAIL_ERR("malloc")
	bzero(page_buffer, file_size);

	len = write(file_fd, page_buffer, file_size);
	if (len < 0)
		FAIL_ERR("write")

	len = lseek(file_fd, 0, SEEK_SET);
	if (len < 0)
		FAIL_ERR("lseek")
	if (len != 0)
		FAIL("len != 0")

	free(page_buffer);
	current_file_size = file_size;
	return (0);
}

static void
run_parent(void)
{
	int connect_socket;
	int status;
	int test_num;
	int test_count;
	int pid;
	size_t desired_file_size = 0;

	const int pagesize = getpagesize();

	struct sendfile_test tests[] = {
 		{ .hdr_length = 0, .offset = 0, .length = 1 },
		{ .hdr_length = 0, .offset = 0, .length = pagesize },
		{ .hdr_length = 0, .offset = 1, .length = 1 },
		{ .hdr_length = 0, .offset = 1, .length = pagesize },
		{ .hdr_length = 0, .offset = pagesize, .length = pagesize },
		{ .hdr_length = 0, .offset = 0, .length = 2*pagesize },
		{ .hdr_length = 0, .offset = 0, .length = 0 },
		{ .hdr_length = 0, .offset = pagesize, .length = 0 },
		{ .hdr_length = 0, .offset = 2*pagesize, .length = 0 },
		{ .hdr_length = 0, .offset = TEST_PAGES*pagesize, .length = 0 },
		{ .hdr_length = 0, .offset = 0, .length = pagesize,
		    .file_size = 1 }
	};

	test_count = sizeof(tests) / sizeof(tests[0]);
	printf("1..%d\n", test_count);

	for (test_num = 1; test_num <= test_count; test_num++) {

		desired_file_size = tests[test_num - 1].file_size;
		if (desired_file_size == 0)
			desired_file_size = TEST_PAGES * pagesize;
		if (write_test_file(desired_file_size) != 0) {
			printf("not ok %d\n", test_num);
			continue;
		}

		pid = fork();
		if (pid == -1) {
			printf("not ok %d\n", test_num);
			continue;
		}

		if (pid == 0)
			run_child();

		usleep(250000);

		if (new_test_socket(&connect_socket) != 0) {
			printf("not ok %d\n", test_num);
			kill(pid, SIGALRM);
			close(connect_socket);
			continue;
		}

		if (send_test(connect_socket, tests[test_num-1]) != 0) {
			printf("not ok %d\n", test_num);
			kill(pid, SIGALRM);
			close(connect_socket);
			continue;
		}

		close(connect_socket);
		if (waitpid(pid, &status, 0) == pid) {
			if (WIFEXITED(status) && WEXITSTATUS(status) == 0)
				printf("%s %d\n", "ok", test_num);
			else
				printf("%s %d\n", "not ok", test_num);
		}
		else {
			printf("not ok %d\n", test_num);
		}
	}
}

static void
cleanup(void)
{

	unlink(path);
}

int
main(int argc, char *argv[])
{

	path[0] = '\0';

	if (argc == 1) {
		snprintf(path, sizeof(path), "sendfile.XXXXXXXXXXXX");
		file_fd = mkstemp(path);
		if (file_fd == -1)
			FAIL_ERR("mkstemp");
	} else if (argc == 2) {
		(void)strlcpy(path, argv[1], sizeof(path));
		file_fd = open(path, O_CREAT | O_TRUNC | O_RDWR, 0600);
		if (file_fd == -1)
			FAIL_ERR("open");
	} else {
		FAIL("usage: sendfile [path]");
	}

	atexit(cleanup);

	run_parent();
	return (0);
}
